src/main/java/bio/terra/cli/apps/DockerAppsRunner.java

```diff
@@ -340,4 +351,47 @@ public static String buildFullCommand(String cmd, List<String> cmdArgs) {
     }
     return fullCommand;
   }
+
+  /**
+   * Build a map of Terra references to use in parsing the CLI command string, and in setting
+   * environment variables in the Docker container.
+   *
+   * <p>The list of references are TERRA_[...] where [...] is the name of a cloud resource. The
+   * cloud resource can be controlled or external.
+   *
+   * <p>e.g. TERRA_MY_BUCKET -> gs://terra-wsm-test-9b7511ab-my-bucket
+   *
+   * @return
+   */
+  private Map<String, String> buildMapOfTerraReferences() {
+    // build a map of reference string -> resolved value
+    Map<String, String> terraReferences = new HashMap<>();
+    workspaceContext
+        .listCloudResources()
+        .forEach(
+            cloudResource ->
+                terraReferences.put(
+                    "TERRA_" + cloudResource.name.toUpperCase(), cloudResource.cloudId));
+
+    return terraReferences;
+  }
+
+  /**
+   * Replace any Terra references in the command string with the resolved values. The references are
+   * case insensitive.
+   *
+   * @param cmd the original command string
+   * @return the modified command string
+   */
+  private String replaceTerraReferences(Map<String, String> terraReferences, String cmd) {
+    // loop through the map entries
+    String modifiedCmd = cmd;
+    for (Map.Entry<String, String> terraReference : terraReferences.entrySet()) {
+      // loop through the map entries, replacing each one in the command string (case insensitive)
+      modifiedCmd =
+          modifiedCmd.replaceAll(
```

Comment:
this doesn't feel very robust, but I think it's fine to go through with this for now with a ticket to follow up later.
There's no escaping. What happens if one of my cloud resource id is a substring of my cloud resource names?

Reply:
Great point. Added a follow-on ticket to address this [PF-459](https://broadworkbench.atlassian.net/browse/PF-459)